tests/cases/fourslash/tsxCompletion14.ts

```diff
@@ -17,12 +17,28 @@
 
 //@Filename: file.tsx
 //// import * as Exp from './exporter';
-//// var x1 = <Exp.Thing /*1*/ />;
-//// var x2 = <Exp.M.SFCComp /*2*/ />;
-//// var x3 = <Exp.Thing /*3*/ ></Exp.Thing>;
-//// var x4 = <Exp.M.SFCComp /*4*/ ></Exp.M.SFCComp>;
+//// var x1  = <Exp.Thing /*1*/ />;
+//// var x2  = <Exp.M.SFCComp /*2*/ />;
+//// var x3  = <Exp.Thing /*3*/ ></Exp.Thing>;
+//// var x4  = <Exp.M.SFCComp /*4*/ ></Exp.M.SFCComp>;
+//// var x5  = <Exp.M.SFCComp><//*5*/>;
+//// var x6  = <Exp.M.SFCComp></Exp./*6*/>;
+//// var x7  = <Exp.M.SFCComp></Exp.M./*7*/>;
+//// var x8  = <Exp.M.SFCComp></Exp.M.SFCComp/*8*/
+//// var x9  = <Exp.M.SFCComp></Exp.M.SFCComp/*9*/>;
+//// var x10 = <Exp.M.SFCComp></      Exp./*10*/>;
+//// var x11 = <Exp.M><//*11*/Exp.M>;
+//// var x11 = <Exp.M.SFCComp></Exp/*12*/>;
 
 verify.completions(
     { marker: ["1", "3"], exact: ["ONE", "TWO"] },
     { marker: ["2", "4"], exact: ["Three", "Four"] },
+    { marker: ["5"], exact: ["Exp.M.SFCComp"] },
+    { marker: ["6"], exact: ["M.SFCComp"] },
+    { marker: ["7"], exact: ["SFCComp"] },
+    { marker: ["8"], exact: [">"] },
+    { marker: ["9"], exact: [""] },
+    { marker: ["10"], exact: ["M.SFCComp"] },
+    { marker: ["11"], exact: [""] },
+    { marker: ["12"], exact: [".M.SFCComp"] },

```

Comment:
@andrewbranch I fixed the other cases and added this and it's failing. I'm not sure if we wanna cover this. Would you confirm if this should also be covered?

Reply:
I removed it. let me know if you want me to cover this case if it makes sense.

```ts
var x12 = <Exp.M.SFCComp></Exp/*1*/>; // Provide leading dot ".M.SFCComp"
```

Reply:
Actually, the general approach here isn’t quite right. The implementation can actually be much simpler than you have it. Completions for a partially-typed entry are not supposed to return the rest of the entry, they’re supposed to return the _whole_ entry, with the `optionalReplacementSpan` set to span of what would get replaced by the _whole_ entry. So in this example, the completion should be `Exp.M.SFCComp` and the `optionalReplacementSpan` should be the span of `Exp` in the closing tag. It doesn’t insert `.M.SFCComp`; it inserts `Exp.M.SFCComp`, overwriting what you’ve already typed. This makes getting the completion text super easy—it’s always the opening tag text—and then you just need to set the `optionalReplacementSpan` in the `getOptionalReplacementSpan` function.

Reply:
Love it. how do we do that?

Reply:
I think you just need to remove all the string slicing (also use `getText()` on `openingTag.tagName` instead of `getFullText()`—you don’t want whitespace), then in `getOptionalReplacementSpan`, you want to return the full closing tag name expression if you are inside a property access in a closing tag. (Right now it will only return the identifier, not the whole property access expression.)

Reply:
Done. Would you review again?
Also while doing this implementation I noticed that we don't have (or I might not have been able to find anything useful) to determine the optional replacement span in fourslash test-cases. I had to do them manually.